### PR TITLE
DR-1798 querying api implementation

### PR DESCRIPTION
**Pull request by tcjiang9**

### CODE REVIEW

File: src/main/java/bio/terra/app/controller/SearchApiController.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/app/controller/SearchApiController.java]: I seem to recall seeing other TDR code where inaccessible objects were filtered out but no error was raised. Either was is fine for now though.

File: src/main/java/bio/terra/app/controller/SearchApiController.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/app/controller/SearchApiController.java]: As a general rule I prefer collecting all errors before reporting, so a user don't have to make multiple calls to find all errors. One way to do this is to use a set difference, e.g. `new HashSet<>(requestIds).removeAll(accessibleIds)`

File: src/main/java/bio/terra/app/controller/SearchApiController.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/app/controller/SearchApiController.java]: it's totally minor but I find the `ResponseEntity` factory methods a bit easier to use. (It was a much bigger deal before java added the diamond operator)
```suggestion
        return ResponseEntity.ok(searchQueryResultModel);
```

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/search/SearchService.java]: Florian has the same comment, maybe this is another ticket to add.

File: src/main/resources/application.properties
Comment (fboulnois) [inline src/main/resources/application.properties]: I don't think we'll want to commit this when we merge, we'll want to overwrite the env var in dev.

File: src/main/java/bio/terra/app/controller/SearchApiController.java
Comment (fboulnois) [inline src/main/java/bio/terra/app/controller/SearchApiController.java]: Nice, this is a great way to achieve this.

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (fboulnois) [inline src/main/java/bio/terra/service/search/SearchService.java]: You may want to check out what I did in my PR, I pulled the `RestHighLevelClient` into the `SearchService` constructor and let it get built automatically. I'm not sure which way is the most ideal.

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (fboulnois) [inline src/main/java/bio/terra/service/search/SearchService.java]: What does the response look like?

**Review by fboulnois — COMMENTED** [review]
A few comments / questions before I approve:

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (tcjiang9) [reply src/main/java/bio/terra/service/search/SearchService.java]: I think neither, we want it to be constructed outside the class and injected anyways

File: src/main/resources/application.properties
Comment (tcjiang9) [reply src/main/resources/application.properties]: yeah this was an accident

File: src/main/java/bio/terra/app/controller/SearchApiController.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/app/controller/SearchApiController.java]: Unnecessary `toString()` (any object in a `String` `+` expression will have `.toString()` called on it automatically).
```suggestion
                        + " on snapshot ids" + inAccessibleIds);
```

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/search/SearchService.java:138]: This is interesting, I noticed that the bulk api Florian is using has both a builder and a non-builder version. Did you happen to notice two styles of apis for this as well?

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/search/SearchService.java]: I'm curious was benefit `@Valid` provides here. Also, is `null` permitted for these arguments?

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/search/SearchService.java]: I'm curious about this too. I would think that this field should be valid JSON, and that this code would be building a JSON array for the response.

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (tcjiang9) [reply src/main/java/bio/terra/service/search/SearchService.java]: I believe if values are not provided, spring will fill these in with their default values, so they will not be null

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (tcjiang9) [reply src/main/java/bio/terra/service/search/SearchService.java]: for a result with two hits it looks like this:

```
{
  "result": "[
        {\"name\":\"John Doe\",\"age\":\"23\",\"background\":\"Born and brought up in California. Engineer by profession. Loves to cook\",\"nationality\":\"British\"}, 
        {\"name\":\"test offset\",\"age\":\"28\",\"background\":\"Born and brought up in California. Engineer by profession. Loves to cook test offset\",\"nationality\":\"British\"}
    ]"
}

result: [array of json blobs as string]

```

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/search/SearchService.java]: I agree, I don't think they can be null. So I would change this code to use `int` instead of `Integer` since the caller is ensuring that they are non-null before they're passed to this layer of the API. Leaving them as `Integer` means we'd have to handle `null` here, which we shouldn't need to do.

Regarding `@Valid`, I think it should be removed unless we know why we're adding it here.

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/search/SearchService.java]: Looks good enough for now, but I think we'll soon want to fix it so it's really JSON and not a string containing JSON. e.g.
```
{
  "result": [
        {"name":"John Doe","age":"23","background":"Born and brought up in California. Engineer by profession. Loves to cook","nationality":"British"}, 
        {"name":"test offset","age":"28","background":"Born and brought up in California. Engineer by profession. Loves to cook test offset","nationality":"British"}
    ]
}
```

Comment (tcjiang9) [reply src/main/java/bio/terra/service/search/SearchService.java]: we would want to change the api to return json instead of a string in that case

**Comment by pshapiro4broad** [comment]
I think it would be either array of object or just object in the open api
yaml. On the Java side, I believe a JsonNode would be rendered as a json
data. But that can wait for a future pr.

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (tcjiang9) [reply src/main/java/bio/terra/service/search/SearchService.java]: I dont think I can change this, the method header is set by spring.

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (tcjiang9) [reply src/main/java/bio/terra/service/search/SearchService.java]: Figured I'd just do it in this pr, the response now looks like

```
{
  "result": [
    {
      "nationality": "British",
      "background": "Born and brought up in California. Engineer by profession. Loves to cook",
      "name": "John Doe",
      "age": "23"
    },
    {
      "nationality": "British",
      "background": "Born and brought up in California. Engineer by profession. Loves to cook test offset",
      "name": "test offset",
      "age": "28"
    }
  ]
}

```

**Review by fboulnois — APPROVED** [review]
lgtm after checkstyles and spotbugs fixes are addressed

File: src/main/java/bio/terra/service/search/SearchService.java
Comment (fboulnois) [inline src/main/java/bio/terra/service/search/SearchService.java]: This can be switched to `SearchException` now too.

**Review by pshapiro4broad — APPROVED** [review]
Looks good, I'm glad you were able to modify it to return JSON directly

**Pull request merged**
